File: src/main/java/com/chattriggers/ctjs/minecraft/libs/renderer/Image.java
package com.chattriggers.ctjs.minecraft.libs.renderer;

import com.chattriggers.ctjs.CTJS;
import com.chattriggers.ctjs.utils.console.Console;
import lombok.Getter;
import lombok.experimental.Accessors;
import net.minecraft.client.renderer.texture.DynamicTexture;
import net.minecraftforge.client.event.RenderGameOverlayEvent;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.fml.common.eventhandler.EventPriority;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;

import javax.imageio.ImageIO;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.net.URL;

/**
 * Used in {@link Renderer#image(String, String)} )}
 */
@Accessors(chain = true)
public class Image {
    private BufferedImage imageToLoad;

    @Getter
    private DynamicTexture texture;

    /**
     * -- GETTER --
     * Gets the image's texture width.
     *
     * @return The image's texture width
     */
    @Getter
    private int textureWidth;

    /**
     * -- GETTER --
     * Gets the image's texture height.
     *
     * @return The image's texture height
     */
    @Getter
    private int textureHeight;

    private Image(BufferedImage image) {
        MinecraftForge.EVENT_BUS.register(this);

        this.imageToLoad = image;
    }

    @SubscribeEvent(priority = EventPriority.HIGH)
    public void render(RenderGameOverlayEvent event) {
        if (imageToLoad != null) {
            texture = new DynamicTexture(imageToLoad);
            this.textureWidth = imageToLoad.getWidth();
            this.textureHeight = imageToLoad.getHeight();

            imageToLoad = null;
        }
    }

    /**
     * Downloads an image to store at the resource name location.
     *
     * @param url          The url to download the image from
     * @return The Image object to allow for method chaining
     */
    public static Image load(String name, String url) {
        File resourceFile = new File(CTJS.getInstance().getAssetsDir(), name);

        try {
            if (resourceFile.exists()) {
                return new Image(ImageIO.read(resourceFile));
            }

            BufferedImage image = ImageIO.read(new URL(url));
            ImageIO.write(image, "png", resourceFile);
            return new Image(image);
        } catch (Exception e) {
            Console.getInstance().out.println("Error loading image " + name + " from url " + url);
            Console.getInstance().printStackTrace(e);
        }

        return null;
    }

    /**
     * Loads an image from a path. Use ./ for the .minecraft folder.
     *
     * @param path The path to load the image from
     * @return The Image object to allow for method chaining
     */
    public static Image load(String path) {
        File file = new File(path);

        try {
            return new Image(ImageIO.read(file));
        } catch (IOException e) {
            Console.getInstance().out.println("Error loading image with path " + path);
            Console.getInstance().printStackTrace(e);
        }

        return null;
    }

    /**
     * Draws the image on screen using the texture width and height.
     *
     * @param x the x position
     * @param y the y position
     * @return The Image object to allow for method chaining
     */
    public Image draw(double x, double y) {
        Renderer.drawImage(this, x, y, this.textureWidth, this.textureHeight);

        return this;
    }

    /**
     * Draws the image on screen with a custom width and height.
     *
     * @param x The x position
     * @param y The y position
     * @param width The image width
     * @param height The image height
     * @return the Image object to allow for method chaining
     */
    public Image draw(double x, double y, double width, double height) {
        Renderer.drawImage(this, x, y, width, height);

        return this;
    }
}

File: src/main/kotlin/com/chattriggers/ctjs/minecraft/libs/renderer/Image.kt
package com.chattriggers.ctjs.minecraft.libs.renderer

import com.chattriggers.ctjs.CTJS
import net.minecraft.client.renderer.texture.DynamicTexture
import net.minecraftforge.client.event.RenderGameOverlayEvent
import net.minecraftforge.common.MinecraftForge
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
import java.awt.image.BufferedImage
import java.io.File
import java.net.HttpURLConnection
import javax.imageio.ImageIO

class Image(var image: BufferedImage?) {
    private lateinit var texture: DynamicTexture
    private val textureWidth = image?.width ?: 0
    private val textureHeight = image?.height ?: 0

    init {
        MinecraftForge.EVENT_BUS.register(this)
        CTJS.images.add(this)
    }

    @Deprecated(
        message = "API is ambiguous, especially when called from JavaScript, and is relative to the assets directory",
        replaceWith = ReplaceWith("Image.fromFile() /* or Image.fromUrl() */"),
    )
    @JvmOverloads
    constructor(name: String, url: String? = null) : this(getBufferedImage(name, url))

    @Deprecated(
        message = "Use static method for consistency",
        replaceWith = ReplaceWith("Image.fromFile()")
    )
    constructor(file: File) : this(ImageIO.read(file))

    fun getTextureWidth(): Int = textureWidth

    fun getTextureHeight(): Int = textureHeight

    fun getTexture(): DynamicTexture {
        if (!::texture.isInitialized) {
            // We're trying to access the texture before initialization. Presumably, the game overlay render event
            // hasn't fired yet so we haven't loaded the texture. Let's hope this is a rendering context!
            try {
                texture = DynamicTexture(image)
                image = null

                MinecraftForge.EVENT_BUS.unregister(this)
            } catch (e: Exception) {
                // Unlucky. This probably wasn't a rendering context.
                println("Trying to bake texture in a non-rendering context.")

                throw e
            }
        }

        return texture
    }

    /**
     * Clears the image from GPU memory and removes its references CT side
     * that way it can be garbage collected if not referenced in js code.
     */
    fun destroy() {
        texture.deleteGlTexture()
        CTJS.images.remove(this)
    }

    @SubscribeEvent
    fun onRender(event: RenderGameOverlayEvent.Pre) {
        if (image != null) {
            texture = DynamicTexture(image)
            image = null

            MinecraftForge.EVENT_BUS.unregister(this)
        }
    }

    @JvmOverloads
    fun draw(
        x: Double, y: Double,
        width: Double = textureWidth.toDouble(),
        height: Double = textureHeight.toDouble()
    ) = apply {
        if (image != null) return@apply

        Renderer.drawImage(this, x, y, width, height)
    }

    @Suppress("DEPRECATION")
    companion object {
        /**
         * Create an image object from a java.io.File object. Throws an exception
         * if the file cannot be found.
         */
        @JvmStatic
        fun fromFile(file: File) =  Image(file)

        /**
         * Create an image object from a file path. Throws an exception
         * if the file cannot be found.
         */
        @JvmStatic
        fun fromFile(file: String) = Image(File(file))

        /**
         * Create an image object from a file path, relative to the assets directory.
         * Throws an exception if the file cannot be found.
         */
        @JvmStatic
        fun fromAsset(name: String) = Image(File(CTJS.assetsDir, name))

        /**
         * Creates an image object from a URL. Throws an exception if an image
         * cannot be created from the URL. Will cache the image in the assets
         */
        @JvmStatic
        @JvmOverloads
        fun fromUrl(url: String, cachedImageName: String? = null): Image {
            if (cachedImageName == null)
                return Image(getImageFromUrl(url))
            return Image(cachedImageName, url)
        }

        private fun getBufferedImage(name: String, url: String? = null): BufferedImage? {
            val resourceFile = File(CTJS.assetsDir, name)

            if (resourceFile.exists())
                return ImageIO.read(resourceFile)

            val image = getImageFromUrl(url!!)
            ImageIO.write(image, "png", resourceFile)
            return image
        }

        private fun getImageFromUrl(url: String): BufferedImage {
            val conn = (CTJS.makeWebRequest(url) as HttpURLConnection).apply {
                requestMethod = "GET"
                doOutput = true
            }

            return ImageIO.read(conn.inputStream)
        }
    }
}
